### ISSUE TITLE
Saving imagesAsPng not working for surf

### ISSUE THREAD

**Issue by Gor-Hakobyan** [matlab2tikz/matlab2tikz#770]
The issue I have is that wheather I set the option imagesAsPng or not, I always get my surf plot in form of data. There is never a separate png generated. If I understood it correctly, that is what should happen. But instead the data are added to the tex file, no matter what the value of 'imagesAsPdf' is. I am trying to save surf figure.

**Comment by egeerardyn** [comment]
The `imageAsPng` function only works for images, i.e. things that are plotted using the `image` or `imagesc` functions, not for "normal" plots such as `surf`.

If you have too much data to plot with `pgfplots`, we have some guidelines on our [wiki](https://github.com/matlab2tikz/matlab2tikz/wiki/TeX-capacity-exceeded,-sorry) how to deal with that. The last point there, is a manual export to PNG, but this is not the `imageAsPng` parameter.

**Comment by Gor-Hakobyan** [comment]
Thank you very much for your answer Egon. In this case I would have a suggestion. There is a way to add extra feature to axis. There is a way to add an extra feature to the tikz. There is a way to add extra code in the beginning and in the end of the generated tex file. But so far known to me there is no way to add anything INSIDE the axis, e.g. Extra addplot command, with which e.g. the externally saved png image can be added to the axes. In case there is indeed no way to do it now, it would be helpful to have it in next versions. I wrote a script that makes the whole image generation with external png automatically based on export_fig and m2t scripts, but it can add the code for \addplot only as extra code, thus, not to the right place. With that feature an automatic generation of images with axes in tex and the plot in png would be possible.

**Comment by egeerardyn** [comment]
We are planning on adding such options in a future version, since this is an often requested feature, e.g. #393. 
Do you have your wrapper scripts somewhere publicly available, because this might interest other users as well?

**Comment by Gor-Hakobyan** [comment]
I have seen a need for such a script and wrote it yesterday. I will work on the code the next days to make it nicer and less error prone in different situations and i'll make it publicly available. But, as I mentioned, the fact that I cannot add \addpot at the right place makes the need of going to the tex file afterwards and change the position of the \addplot per hand. Still, the script does the job (at least for 2d surf images), and might be useful for other users.

**Comment by egeerardyn** [comment]
Depending on how complex everything in your script is, another option could be to integrate this directly into `matlab2tikz`. That would certainly be welcome in a pull request.

In any case: you can expect the proper interface to be implemented on our side such that code can be inserted at strategic places. I expect the implementation shouldn't take too long, but don't expect an implementation before October.

**Comment by okomarov** [comment]
@Gor-Hakobyan Also, be aware that you can set the axis labels (tick labels included) to use latex fonts. So, if you exported the whole figure in .eps, then your compiled figure would be consistent with your latex document.

**Comment by Gor-Hakobyan** [comment]
@egeerardyn well, both solutions will do the job and be useful. I think by the end of the week I could send you the script. I use export_fig for creating the png image, but there are certainly other ways for doing it.

@okomarov thank you for pointing out the option with eps. I have tried to go that way, but I had a problem with including the eps images, while using pdflatex. There is a package "epstopdf", which is supposed to solve this issue. I have never tried it out. Setting the text interpreter in latex to matlab is not an issue, but what about the size of the fonts? Are they also adjusted according to your latex script fonts?

Well, initially the reason I couldn't live with the images generated in latex with the help of m2t was the fact that matlab2tikz was setting the option "shader flat", which leads to an output which is considerably different than that of matlab. The point is that this option averages the neighboring values for plotting the surf. Mr. Feuersänger pointed out, that this issue can be solved by putting "shader flat corner". This was the initial reason. However, exporting the surf images as png might be handy for surf figures with a lot of data points. There are surely other ways of doing it also, as @egeerardyn pointed out in his comment above.

@egeerardyn the feature of setting "shader flat", which is the acronym for "shader flat mean", is in my opinion an unwanted behaviour, resulting in an output different from the original one in matlab. It took me days to find it out.

**Comment by okomarov** [comment]
Be aware that png is a raster image. This means that you lose definition if you zoom in.

It would help us if you raised a new ticket with the shader issue. We would need a minimal matlab example that produces the problem together with the generated latex code.

Eps with latex labels should preserve size specified in the tex document.

Sent from my iPhone

On 11 Sep 2015, at 10:42, "Gor-Hakobyan" <notifications@github.com<mailto:notifications@github.com>> wrote:


@egeerardyn<https://github.com/egeerardyn> well, both solutions will do the job and be useful. I think by the end of the week I could send you the script. I use export_fig for creating the png image, but there are certainly other ways for doing it.

@okomarov<https://github.com/okomarov> thank you for pointing out the option with eps. I have tried to go that way, but I had a problem with including the eps images, while using pdflatex. There is a package "epstopdf", which is supposed to solve this issue. I have never tried it out. Setting the text interpreter in latex to matlab is not an issue, but what about the size of the fonts? Are they also adjusted according to your latex script fonts?

Well, initially the reason I couldn't live with the images generated in latex with the help of m2t was the fact that matlab2tikz was setting the option "shader flat", which leads to an output which is considerably different than that of matlab. The point is that this option averages the neighboring values for plotting the surf. Mr. Feuersänger pointed out, that this issue can be solved by putting "shader flat corner". This was the initial reason. However, exporting the surf images as png might be handy for surf figures with a lot of data points. There are surely other ways of doing it also, as @egeerardyn<https://github.com/egeerardyn> pointed out in his comment above.

@egeerardyn<https://github.com/egeerardyn> the feature of setting "shader flat", which is the acronym for "shader flat mean", is in my opinion an unwanted behaviour, resulting in an output different from the original one in matlab. It took me days to find it out.

**Comment by egeerardyn** [comment]
@okomarov In some cases a bitmap is a better idea than vector (e.g. in my department it's quite common to show DFT spectra with thousands of points of random signals. Simplification algorithms are pretty much useless due to the randomness, but plotting each point is just way too expensive (for compilation time (or buffer restraints) if using `pgfplots` and file size if using native PDF/EPS export (some venues limit the paper file size)) unless the data is exported as a PNG image, e.g. (from a draft paper I co-authored with a colleague of mine)

![check](https://cloud.githubusercontent.com/assets/177360/9814307/235db8e0-588d-11e5-9684-94684f22c850.png)

For 3D plots, I think there are also enough situations where it is not feasible for most users to downsample their data to a suitable resolution. In those cases, exporting the data part to PNG is a quick workaround that I think is very worthwhile.

For the shader, I agree completely: we should look at that.

@Gor-Hakobyan The `epstopdf` LaTeX package is very easy to use. Just include it in your document and make sure you call `pdflatex` with the `-shell-escape`  option and it should work very well. In some TeX distributions and editors it is also enabled behind your back: it just works that well.

But if you can provide an MWE for the shader problems you see, we can certainly have a look and see if we can improve the behavior.

**Comment by Gor-Hakobyan** [comment]
@egeerardyn I just wanted to create new issue about the shader option with a MWE, but couldn't find a way to attach some files not in an image format. I would like to send the matlab data, script, pdf file showing the issue, and maybe the .tex files. I can pack it into zip and change the extension to .png. Is this the way or you know a more elegant solution?

**Comment by egeerardyn** [comment]
@Gor-Hakobyan: normally, we can work with just a MATLAB script (with the data integrated, often this can be simplified to a small subset). To make your point more obvious, we also appreciate a side-by-side screenshot of the compiled PDF from `matlab2tikz` and the original MATLAB figure. We don't have any need for full TeX file: we can just generate them from MATLAB code.

If your data is too complex to put into the script yourself, you can also try to save it to an [M file](http://nl.mathworks.com/help/matlab/ref/matlab.io.savevariablestoscript.html), a [ASCII data](http://nl.mathworks.com/help/matlab/ref/save.html?searchHighlight=save) file which you can upload as a comment or even a [gist](https://gist.github.com).

**Comment by Gor-Hakobyan** [comment]
@egeerardyn #772 I formulated that isssue and posted the MWE as a matlab file to the comments. Looks pretty bad, taking the whole place. :)

**Comment by Gor-Hakobyan** [comment]
@egeerardyn Here is a small script which creates tex files just like ```matlab2tikz``` but with data saved in png format. This might be handy if the data to plot are huge and cannot be handled with other means (example above). Or just in case nothing delivers the desired result. :) So far it is not designed for 3D images, but I could probably add it soon if I don't find a better way of bringing 3D matlab images to latex.

The script uses ```matlab2tikz``` and ```export_fig```. For further infos see the description of the function.

```
% Automatical creation of tikz graphics with png images with the
%help of matlab2tikz and export_fig scripts.
%   The passed parameter should be the figure handle, from which the image
%   has to be created
% Input parameters:
%   - fig_handle is the handle to the figure to be extracted: default:gcf
%   - fig_name is the name both the png and the .tex file are going to have
%   - fig_height is the height of the object: default 2.5in
%   - fig_width is the width of the object: default 0.4\textwidth
% IMPORTANT: after generation of the data you have to move the command
% \addplot into the axis{}
% This script uses export_fig and matlab2tikz packages
function [ output_args ] = matlab2tikz_png(fig_handle, fig_name, fig_height, fig_width)
% Creating tikz images with graphs saved as external pngs
% Gor Hakobyan. 13.09.2015

temp_fig=figure;
set(temp_fig, 'Position', get(fig_handle, 'Position'))
warning off
copyobj(allchild(fig_handle), temp_fig)
warning on
cla
if(~exist('fig_height', 'var'))
    fig_height='2.5in';
elseif(~ischar(fig_height))
    error('Please specify a valid height in string format')
end
if(~exist('fig_width', 'var'))
    fig_width='0.4\textwidth';
elseif(~ischar(fig_width))
    error('Please specify a valid width in string format')
end

allAxesInFigure = findall(fig_handle,'type','axes'); % !The case of multiple axes not covered yet
set(allAxesInFigure, 'Visible', 'off')
export_fig([fig_name, '.png'], '-q101', '-a1', fig_handle)

set(allAxesInFigure, 'Visible', 'on')
% axes_child=get(allAxesInFigure, '')
% set(allchild(allAxesInFigure), 'Visible', 'off')

% Get the limits of axis in order to position the png correctly
x_lim=get(allAxesInFigure, 'Xlim');
y_lim=get(allAxesInFigure, 'Ylim');
ext_add_plot=['\addplot [forget plot] graphics [xmin=', num2str(x_lim(1)), ...
    ',xmax=', num2str(x_lim(2)),',ymin=',num2str(y_lim(1)),',ymax=', ...
    num2str(y_lim(2)),'] {fig/Grafik/tikz_png/',fig_name, '.png};'];
matlab2tikz([fig_name, '.tex'], 'height', fig_height, 'width', fig_width,...
    'standalone',false,'imagesAsPng', true, ...
    'extraAxisOptions', 'ylabel near ticks, xlabel near ticks', ...
    'extraCode', ext_add_plot, 'figurehandle',temp_fig)

output_args='ok';
set(allAxesInFigure, 'Visible', 'on')
close(temp_fig)
end
```

**Comment by egeerardyn** [comment]
@Gor-Hakobyan If I understand your script correctly, you only use `pgfplots` to plot the axes and let `export_fig` handle the actual data of all plots.

That can be a nice start for some people, indeed.

I will try to have a look what implications this has on our side, but I don't think this should be a problem on our side. I'm already contemplating whether we can extend this to 3D (that would be really nice) and whether we could provide an easier to use mechanism to rasterize only certain subplots or parts of a plot.

**Comment by Gor-Hakobyan** [comment]
@egeerardyn I was thinking about extending it to 3D, since it would make the script much more useful. I have seen a couple of ideas how to implement it, but  had unfortunately no time to try it out. Hopefully I find time for it the next days.

By the way, saving the data as png is not the only way. Instead, also a vector graphics format, such as pdf, can be used. I don't know whether there is a use case, but I guess it might make sense in certain situations.

However, be avare that the vector graphics output, generated by painters renderer in matlab, has each rectangle of surf plotted as two triangles, which somehow leads to a dashed line displayed between those in e.g. Pdf reader. This is a very nasty issue, to which I found no solution but not using painters renderer of matlab.

**Comment by Druidefix** [comment]
As @egeerardyn pointed out, there are use cases which do not allow a down sampling of the data to plot. I was looking for a way to force matlab2tikz to create a pixel graphic file instead of just saving thousands of coordinate points. If you plan to enhance matlab2tikz this would be create, so far I wrote a function to cope with this issue.

I distinguish three cases, for which I have separate creating rules:

1) A normal plot of lots of data as depicted in the image above. This is just a simple

```
% hide the axis environment, because this is set up using pgfplots later on
axis off
% save the actual plot as a png
print -dpng surf2D -r400
% trim all the white space which matlab adds to the image
system('convert -trim surf2D.png surf2D.png');
% set the white background to transparent
system('convert -transparent white surf2D.png surf2D.png');
```

Afterwards I output the code using \addplot graphics with the correct xmin, xmax, ymin and ymax values. Matlab always adds whitespace around the figure, so I use imagemagick under Linux to remove the white space. Unfortunately, this part is depending on the system call and the availability of imagemagick.

2) To plot a 2D image of a large matrix I use the following code

```
imagesc(x, y, amp)
% imagesc starts the y dimension on top left instead of bottom left
% -> use normal direction
set(gca, 'YDir', 'normal');
matlab2tikz('imagesc.tikz', 'noSize', true, 'showInfo', false, 'showWarnings', false)
```

By using matlab2tikz the created png file has one pixel value for each matrix cell which is better suited than just creating the figure with the print function.

3) For a 3D plot I use the approach presented by [Jake](https://tex.stackexchange.com/questions/52987/), but his script to determine the coordinates does not work. So I use approach 1) and output a template in which I input the correct values by hand using Gimp.


In summary, it would be great if there is a matlab2tikz option like 'create_rasterimage' which can be set to true to get matlab2tikz use the \addplot graphics option. My approach is just a workaround, I have to input the data to plot, then the plot is set up and exported. It does not use a given figure and therefore labels are not converted. For me the important part is that the correct image and axis information are created, the rest is done by hand and adepted as I need it.

I only presented the idea of my approach, if you are interested in the basic implementation, I could share the function code.

**Comment by egeerardyn** [comment]
@Druidefix Certainly your first case is something that I see in this feature. The second case, I think, we handle already. If we don't, it's a bug.
The third case is an extension of the first, but it's a tough challenge, I think.

But I really do see potential for including this in `m2t`, it's just that this is not a short-term high-priority thing.

The dependency on `imagemagick` is something we should be very explicit about if we would include this in `m2t`. I think the clearest/easiest approach for such things is to make a tiny MATLAB wrapper around such external dependencies where we can do some error checking and allowing the user to specify a path to the right executable (from experience, I know this can be a pain in MATLAB since not all the system paths are carried over automatically and on Windows it's actually very uncommon to adapt the PATH such that executables are reachable).

Your ideas make me wonder whether it would be possible to use the Image Processing toolbox to do some parts of the job. Maybe it would even be possible to map pixels to coordinates by exporting a test image analyzing it with that toolbox. That could alleviate exporting 3D images.

**Comment by Druidefix** [comment]
After thinking about my work flow and talking to other students, I would say that the cases are somehow connected. Until now you definitely have to know, that if you want to convert a large 2D matrix plot to TikZ, you have to choose imagesc to plot the matrix. Other colleagues use different Matlab functions to plot, e.g. surf or pcolor. With those functions creating a png is not possible as far as I know. You must be aware of the imagesc way.

So I would expect m2t to behave something like this:
1) I convert a image regardless if it's case 1), 2) or 3).
2) m2t checks, if the number of points is above a threshold and then sets a variable to true.
3) For case 2) it uses the imagesc with axis reverse to convert. For the other cases it uses a different approach.

Nevertheless, I can force m2t by setting a variable to create or to not create an image.

If there is a way how I could support the m2t team to introduce the feature, please let me know.

**Comment by mhham** [comment]
Hello there,
Any news on this feature ?
Or does anyone know a simple package/program that exports the content of a matlab figure as png and the axes as tikz ?
Cheers !

**Comment by Druidefix** [comment]
For the procedure I mentioned above, there exists a function written from my college for just this specific case. You can have a look at https://github.com/Fiech/surf3tikz
